src/parser/shared/modules/BuffCountGraph.tsx

```diff
@@ -0,0 +1,440 @@
+  /** Trackers for the rule lines to graph */
+  ruleLineTrackers: GraphedSpellInternalTracker[];
+  /** Rule line trackers indexed by spellId */
+  ruleLineTrackerLookup: { [key: number]: GraphedSpellInternalTracker[] };
+  /** Data handed to vega for graphing */
+  graphData: GraphData[];
+  /** Timestamp of the last data points added - used to avoid adding duplicate data on same timestamp */
+  lastTimestamp: number;
+
+  protected constructor(options: Options) {
+    super(options);
+
+    // create internal tracking objs
+    this.buffTrackers = this.buffSpecs().map((spec) => this._convertToInternalTracker(spec));
+    this.buffTrackerLookup = this._generateLookup(this.buffTrackers);
+    this.ruleLineTrackers = this.castRuleSpecs().map((spec) =>
+      this._convertToInternalTracker(spec),
+    );
+    this.ruleLineTrackerLookup = this._generateLookup(this.ruleLineTrackers);
+    // create empty data list
+    this.graphData = [];
+    this.lastTimestamp = this.owner.fight.start_time; // timestamp of last item added to data
+
+    // setup listeners - spell list needs to filtered through a Set to remove duplicates
+    const buffSpellList = [...new Set(this.buffTrackers.flatMap((spec) => spec.spells))];
+    this.addEventListener(
+      Events.applybuff.by(SELECTED_PLAYER).spell(buffSpellList),
+      this.onBuffApplied,
+    );
+    this.addEventListener(
+      Events.removebuff.by(SELECTED_PLAYER).spell(buffSpellList),
+      this.onBuffRemoved,
+    );
+    this.addEventListener(
+      Events.applydebuff.by(SELECTED_PLAYER).spell(buffSpellList),
+      this.onBuffApplied,
+    );
+    this.addEventListener(
+      Events.removedebuff.by(SELECTED_PLAYER).spell(buffSpellList),
+      this.onBuffRemoved,
+    );
+
+    const castSpellList = [...new Set(this.ruleLineTrackers.flatMap((spec) => spec.spells))];
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(castSpellList), this.onCast);
+
+    this.addEventListener(Events.fightend, this.onFightEnd);
+  }
+
+  /**
+   * Implement to return the buffs to be graphed.
+   * This is a function instead of a field in the constructor so implementer can filter the specs
+   * based on player talents / legendaries / covenant.
+   */
+  abstract buffSpecs(): GraphedSpellSpec[];
+
+  /**
+   * Override to return casts to be graphed.
+   * This is a function instead of a field in the constructor so implementer can filter the specs
+   * based on player talents / legendaries / covenant.
+   */
+  castRuleSpecs(): GraphedSpellSpec[] {
+    return [];
+  }
+
+  /**
+   * Adds a vertical rule line for the given tracker at the given timestamp
+   * @param trackerName the name of the tracker this rule line is for -
+   *     will determine the line's color and its name in the legend
+   * @param timestamp the timestamp to add the rule
+   */
+  addRuleLine(trackerName: string, timestamp: number) {
+    const tracker = this.ruleLineTrackers.find((t) => t.name === trackerName);
```

Comment:
Would making a map with the keys the name bet quicker? reducing this O(n) to O(1)?